### PR TITLE
Rank students function implementation and tests

### LINKED ISSUE #7: Function - Summarise the grades by students
Purpose:
Calculate the average grade for all students and provide the ranking for the course selected or the whole program.
Lecturers can find students that are struggling with the courses

Input - a specified course with a default option for the entire program
Output - student’s ranking

### PR DESCRIPTION
**Pull request by jraza19**

### CODE REVIEW

File: R/frgtracker.R [hunk 1]
```diff
@@ -82,32 +81,88 @@ rank_courses <- function(method= c("method", "median", "lst-quantile", "3rd-quan
...
 #'
 #' @examples
-#' rank_students()
-#' rank_students(course_id = "511", n = 3, ascending = TRUE)
-rank_students <- function(course_id = "all", n = 3, ascending = FALSE) {
-  print("NULL")
+#' df <- data.frame(course_id  = c(rep("511", 4)),
+#' student_id =  c("tom", "tiff", "mike", "joel"),
+#' grade = c(90, 80, 70, 67)
+#'
```
Comment (tdkhanhvu) [inline R/frgtracker.R]: extra space

File: R/frgtracker.R [hunk 2]
```diff
@@ -82,32 +81,88 @@ rank_courses <- function(method= c("method", "median", "lst-quantile", "3rd-quan
...
+  if (!(courseid %in% c(unique(df$course_id), "all"))){
+    stop("This course is currently not a part of the courses list")
+  }
+
+  slicer <- ifelse(ascending, dplyr::slice_min, dplyr::slice_max)
+
+  temp_df <- df
+
+  if (courseid != "all"){
+    temp_df <-	df %>%
```
Comment (tdkhanhvu) [inline R/frgtracker.R]: extra spaces

File: tests/testthat/test-frgtracker.R [hunk 3]
```diff
@@ -19,6 +16,86 @@ test_that("multiplication works", {
 
 #function5 start
```
Comment (tdkhanhvu) [inline tests/testthat/test-frgtracker.R]: can you rename the comment here to reflect the function name?

File: tests/testthat/test-frgtracker.R [hunk 4]
```diff
@@ -19,6 +16,86 @@ test_that("multiplication works", {
...
+  df <- data.frame (course_id  = c(rep("511", 4), rep("522", 4)),
+                    student_id =  c("tom", "tiff","mike","joel",
+                                    "tom","tiff","mike","joel"),
+                    grade = c(90, 70, 60, 50, 95, 91, 85, 80)
+  )
+  df
+}
+
+
+generate_fake_dataframe_output_511_course <- function(){
```
Comment (tdkhanhvu) [inline tests/testthat/test-frgtracker.R]: for the 3 functions to generate the fake data, you can merge them into 1.

Parameter will be the grades to pass in. The rest should be the same

File: tests/testthat/test-frgtracker.R [hunk 5]
```diff
@@ -19,6 +16,86 @@ test_that("multiplication works", {
...
+
+generate_fake_matrix <- function(){
+  matrix_ex <- matrix(data = c(c(rep("511",4), rep("522", 4)),
+                               c("tom", "tiff","mike","joel","tom",
+                                 "tiff","mike","joel"),
+                               c(84.66, 88.34, 87.66, 90.82, 95.52,
+                                 87.92, 88.92, 92.80)),
+                      nrow = 8,
+                      ncol = 3)
+}
```
Comment (tdkhanhvu) [inline tests/testthat/test-frgtracker.R:57]: I don't think you return the matrix here?

File: tests/testthat/test-frgtracker.R [hunk 6]
```diff
@@ -19,6 +16,86 @@ test_that("multiplication works", {
...
+  expect_false(any(temp_df$grade > 100))
+})
+
+test_that("NAs should be dropped", {
+  temp_df = generate_fake_dataframe()
+
+  expect_false(is.null(rank_students(df = temp_df)))
+})
+
 #function5 end
```
Comment (tdkhanhvu) [inline tests/testthat/test-frgtracker.R]: can you rename the function name here as well?

**Review by tdkhanhvu — COMMENTED** [review]
It looks good to me in general. Only a few small comments

File: R/frgtracker.R (hunk 1 above)
Comment (jraza19) [reply R/frgtracker.R]: done!

File: R/frgtracker.R (hunk 2 above)
Comment (jraza19) [reply R/frgtracker.R]: done!

File: tests/testthat/test-frgtracker.R (hunk 3 above)
Comment (jraza19) [reply tests/testthat/test-frgtracker.R]: done!

File: tests/testthat/test-frgtracker.R (hunk 4 above)
Comment (jraza19) [reply tests/testthat/test-frgtracker.R]: good catch :) thanks for this! done!

File: tests/testthat/test-frgtracker.R (hunk 5 above)
Comment (jraza19) [reply tests/testthat/test-frgtracker.R:57]: i added the name at the end.

File: tests/testthat/test-frgtracker.R (hunk 6 above)
Comment (jraza19) [reply tests/testthat/test-frgtracker.R]: done!

**Review by jianructose — APPROVED** [review]
nice!

**Comment by jraza19** [comment]
New PR #16 related to this function as there are too many merge conflicts with this PR

**Pull request closed without merging**